lib/src/common/cache/cache.dart

```diff
@@ -294,6 +372,37 @@ class Cache<TIdentifier, TValue> extends Object with Disposable {
     return new Future.value();
   }
 
+  /// Returns `true` and calls [callback] if there is a cached, unreleased
+  /// [TValue] associated with [id]; returns `false` otherwise.
+  ///
+  /// This does not perform a [get] or [getAsync], and as a result, will not
+  /// affect retention or removal of a [TIdentifier][TValue] pair from the cache.
+  /// If any callbacks are in fight on removal of [TIdentifier] [TValue] pair
+  /// [didRemove] will not event until callback has completed.
+  ///
+  /// If the [Cache] [isOrWillBeDisposed] then a [StateError] is thrown.
+  Future<bool> applyToItem(
+      TIdentifier id, Future<Null> callback(Future<TValue> value)) {
+    _log.finest('applyToItem id: $id');
+    _throwWhenDisposed('applyToItem');
+    if (_isReleased[id] != false) {
+      return new Future.value(false);
+    }
+    final callbackFuture = callback(_cache[id]);
+    if (callbackFuture is Future<Null>) {
+      _applyToItemCallBacks.putIfAbsent(id, () => <Future<Null>>[]);
+      _applyToItemCallBacks[id].add(callbackFuture);
+      callbackFuture.then((_) {
+        _applyToItemCallBacks[id].remove(id);
+        if (_applyToItemCallBacks[id].isEmpty) {
+          _applyToItemCallBacks.remove(id);
+        }
```

Comment:
Is there a way you can test this?

Reply:
I think [this test](https://github.com/Workiva/w_common/pull/90/files#diff-71e7ef0509871bfd43a03a17b3e52771R452) should cover it. The more I think about it the less worried about a concurrent modification issue, there can only be one thing actually executing at one time.

Reply:
The problem isn't two things executing at the same time, it's mutating a list while it's being iterated.

This case looks pretty safe from that case, though. If you want more peace of mind, a good way I've found to prevent concurrent modification errors is to make a copy of the iterable first, like so:
```dart
await Future.wait(_applyToItemCallBacks[id].toList());
```

---

Other gotchas around `Future.wait` to look out for that I found while thinking about this case:

Using `Future.wait` with a `.map` in which modifications can happen during iteration can be risky, especially in Dart 2 once async functions run their first part synchronously:
```dart
main() {
  var values = [1, 2];
  Future.wait(values.map((value) {
    return new Future.sync(() {
      values.remove(value); // Concurrent modification during iteration: Instance of 'MappedListIterable'.
    });
  }));
}
```
```dart
main() {
  var values = [1, 2];
  Future.wait(values.map((value) {
    values.remove(value); // Concurrent modification during iteration: Instance of 'MappedListIterable'.
    return new Future(() {});
  }));
}
```
```dart
// Dart 2
main() {
  var values = [1, 2];
  Future.wait(values.map((value) async {
    values.remove(value); // Concurrent modification during iteration: Instance of 'MappedListIterable'.
    await new Future(() {});
  }));
}
```

Reply:
Yeah what I meant to say was that the only way we would get mutation of a list while it's being iterated with the current structure is if we had multiple things executing at one time.